File: TP Final/Haskell/Color.hs
module Color where

data Color = RGB Int Int Int
     deriving (Eq, Show)
     {- INV.REP.: los 3 números están entre 0 y 254 -}

renderColor :: Color -> String
renderColor (RGB r g b) = "rgb(" ++ show r ++ "," ++ show g ++ "," ++ show b ++ ")"

black, white                   :: Color
red, green, blue               :: Color
yellow, orange, purple         :: Color
pink, cyan, brown              :: Color
green1, green2, green3, green4 :: Color
brown1, brown2                 :: Color
black  = RGB   0   0   0
white  = RGB 255 255 255
red    = RGB 255   0   0
green  = RGB   0 255   0
blue   = RGB   0   0 255
yellow = RGB 255 255   0
orange = RGB 255 180   0
purple = RGB 180   0 255
pink   = RGB 255   0 255
cyan   = RGB   0 255 255
brown  = RGB 120  60  60 
green1 = RGB   0 129   0
green2 = RGB   0 166   0
green3 = RGB   0 194   0
green4 = RGB 170 212   0
brown1 = RGB 170  83   0
brown2 = RGB 170 136   0

File: TP Final/Color.cpp
#include <iostream>
using namespace std;

#include "Color.h"

struct ColorStr { 
  // INV.REP.: 0<=red<=255, 0<=green<=255, 0<=blue<=255, 
  int red;
  int green;
  int blue;
};

// PRECOND: 0<=r<=255, 0<=g<=255, 0<=b<=255, 
Color rgb(int r, int g, int b) {
  ColorStr* c = new ColorStr;
  c->red   = r;
  c->green = g;
  c->blue  = b;
  return c;
}

bool equalColor(Color c1, Color c2) {
  return(c1->red  ==c2->red
      && c1->green==c2->green
      && c1->blue ==c2->blue);
}

void renderChannel(int c) {
  //if (c<10)  { cout << " "; }
  //if (c<100) { cout << " "; }
  cout << c;
}

int expectedChannelLength(int c) { return (c<10?1:(c<100?2:3)); }
int expectedColorLength(Color c) {
  return (7 + expectedChannelLength(c->red)
            + expectedChannelLength(c->green)
            + expectedChannelLength(c->blue));
}

void renderBlanks(int n) {
  for(int i=1;i<=n;i++) { cout << " "; }
}

void RenderColor(Color c, int minLen) {
	// OBS: 
	//  El ancho esperado está entre 10 y 16. 
	//  Si minLen es menor que el ancho esperado, se obtiene ese ancho.
	//  Si es mayor o igual, se obtiene minLen.
  cout << "rgb(";
  renderChannel(c->red);
  cout << ","; 
  renderChannel(c->green);
  cout << ",";
  renderChannel(c->blue);
  cout << ")";
  renderBlanks(max(0,minLen-expectedColorLength(c)));
}
